### PR TITLE
Add some shape decomps (t, transpose, rot90, stack)

### PR DESCRIPTION
**Pull request by Chillee**
Also fixes xlogy (turns out the only thing it was missing was a type cast annotation! nice!)

I also renamed `canonicalize_idx` => `canonicalize_dim` (to align with `canonicalize_dims`) and fixed a bug in it (cc: @mruberry)

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -1152,3 +1157,69 @@ def cudnn_batch_norm_backward(
...
+@register_decomposition(aten.rot90.default)
+def rot90(self: Tensor, k: int = 1, dims: List[int] = [0, 1]) -> Tensor:
+    total_dims = self.dim()
+    total_rot_dims = len(dims)
+    assert total_rot_dims == 2, f"expected total rotation dims == 2, but got dims = {total_rot_dims}"
+    assert total_dims >= 2, f"expected total dims >= 2, but got total dims = {total_dims}"
+    assert dims[0] != dims[1] and abs(dims[0] - dims[1]) != total_dims, f"expected rotation dims to be different, but got dim0 = {dims[0]} and dim1 = {dims[1]}"
+    assert dims[0] < total_dims and dims[0] >= -total_dims, f"Rotation dim0 out of range, dim0 = {dims[0]}"
+    assert dims[1] < total_dims and dims[1] >= -total_dims, f"Rotation dim1 out of range, dim1 = {dims[1]}"
+    k = (4 + (k % 4)) % 4
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:1171]: what's this? Modulo of a negative number is what you'd expect (so k=-1 will translate to 3)

Comment (Chillee) [reply torch/_decomp/decompositions.py:1171]: uh... too literal translation of C++ code to Python :P

File: torch/_prims/utils.py [hunk 2]
```diff
@@ -237,6 +237,8 @@ def canonicalize_idx(rank: int, idx: int) -> int:
 
     if idx < 0:
         _idx = idx + _rank
+    else:
```
Comment (mruberry) [inline torch/_prims/utils.py:240]: Nice

**Comment by Chillee** [comment]
Could I get an approval on this? (@mruberry, @ngimel)

**Review by mruberry — APPROVED** [review]
Stamp

**Comment by Chillee** [comment]
(passed all tests on a previous commit and I just fixed a merge issue, so hopefully good) :)

**Pull request closed without merging**